### PR TITLE
Prevent hang in git cat-file if repository is not a valid repository and other fixes

### LINKED ISSUE #9271: iowait and memory consumption are skyrocketing regularly
- Gitea version (or commit ref): 1.10.0
- Git version: 2.16.4
- Operating system: openSUSE Leap 15.1
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL (MariaDB)
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description

As written earlier [in the Discourse forum](https://discourse.gitea.io/t/gitea-suddenly-consumes-all-available-memory/1465/2), it regularly happens that gitea makes my server's iowait and memory consumption skyrocket, causing issues for all other services running within the same VM. Usually only killing the whole gitea service makes the server recover.
While I couldn't find any direct cause by parsing the logs or the PPROF data, I just discovered in the admin web interface (tab Monitoring) that the **GetDiffRange** operation seems to be responsible. It seems to get stuck at a single repo of just 138MiB size, clogging IO and making gitea consume more and more memory. `iotop` shows that gitea creates a lot of I/O, although I'm not sure whether swapping operations are allocated to the process causing them as well.
After this issue first occured, I limited the gitea systemd service to 600MiB and gitea is currently using all of it. But apparently this limit isn't working as the memory+swap usage increased by more than 3GiB during the GetDiffRange operation.

What does this **GetDiffRange** operation do? I wasn't able to find information on that.

Luckily I was able to allocate an additional 1GiB of memory to my VM, giving me the chance to let the operation run. So far it's running for more than 30 minutes though.

Any ideas about that operation, why it's getting stuck or any additional debug information I could provide?

### LINKED ISSUE #14734: It takes a long time to open a big pull request page
Gitea version 1.14.0+dev-449-ge0c753e77 built with GNU Make 4.3, go1.15.6 : bindata, timetzdata, sqlite, sqlite_unlock_notify

The gitea server is deployed locally. I made a pull request and changed more than 100 files and more than 1,000 lines. It took me nearly two minutes to open this pull request page.
I looked at the log, mainly because these two API return very slowly, both of which are about one minute
Are there any configuration items that can improve the speed here?
image

### LINKED ISSUE #16113: Task running since 2 weeks
- Gitea version (or commit ref): 1.14.2
- Git version:
- Operating system: Ubuntu 20.04
  
  
  
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Log gist:

## Description

Just found today the task, which is running since 2 weeks. There is no system git process.

And it is not possible to remove it.

## Screenshots

image

image

### LINKED ISSUE #17943: Clicking on a feature branch name from a dashboard entry directs to the wrong url (404)
### Gitea Version

1.16.0+dev-670-gc7e23401a

### Git Version

2.34

### Operating System

windows

### How are you running Gitea?

windows service 

### Database

MySQL

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Description

reports on the dashboard indicate that some parts are clickable 

**eeyrjmr** pushed to **some_branch** at **eeyrjmr/test_issues**

However the branch url points to  **https://try.gitea.io/some_branch**  instead of **https://try.gitea.io/eeyrjmr/test_issues/src/branch/some_branch**   ie the base url does not consider the user/organisation + repository 

Annotation 2021-12-10 000603
Annotation 2021-12-10 000627

### Screenshots

_No response_

### LINKED ISSUE #17983: Various actions on pull requests are broken
### Gitea Version

1.16.0+dev-687-g0981ec30c

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

try.gitea.io

### Database

_No response_

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Description

Various actions on pull requests are broken because the paths point to `.../pulls/...` instead of  `.../issues/...` and these actions are only supported on the issues router:

* toggling WIP status
* subscribe/unsubscribe to notifications
* time tracker
* add/remove issue dependencies
* lock/unlock conversation

A regression from PR #17551 I believe. Same for #17943.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by zeripath**
This PR contains multiple fixes. The most important of which is:

* Prevent hang in git cat-file if the repository is not a valid repository 
    
    Unfortunately it appears that if git cat-file is run in an invalid
    repository it will hang until stdin is closed. This will result in
    deadlocked /pulls pages and dangling git cat-file calls if a broken
    repository is tried to be reviewed or pulls exists for a broken
    repository.

    Fix #14734
    Fix #9271
    Fix #16113

Otherwise there are a few small other fixes included which this PR was initially intending to fix:

* Fix panic on partial compares due to missing PullRequestWorkInProgressPrefixes
* Fix links on pulls pages  due to regression from #17551 - by making most /issues routes match /pulls too - Fix #17983
* Fix links on feeds pages due to another regression from #17551 but also fix issue with syncing tags - Fix #17943
* Add missing locale entries for oauth group claims
* Prevent NPEs if ColorFormat is called on nil users, repos or teams.

### CODE REVIEW

**Review by lunny — APPROVED** [review]

**Comment by KN4CK3R** [comment]
Issue/Pull links are still not working in the news feed:
grafik

File: templates/user/dashboard/feeds.tmpl [hunk 1]
```diff
@@ -17,71 +17,67 @@
...
 						{{else if eq .GetOpType 7}}
 							{{ $index := index .GetIssueInfos 0}}
-							{{$.i18n.Tr "action.create_pull_request" (.GetRepoLink|Escape) $index (.ShortRepoPath|Escape) | Str2html}}
+							{{$.i18n.Tr "action.create_pull_request" ((printf "%s/pulls/%s" .GetRepoLink $index) |Escape) $index (.ShortRepoPath|Escape) | Str2html}}
 						{{else if eq .GetOpType 8}}
 							{{$.i18n.Tr "action.transfer_repo" .GetContent (.GetRepoLink|Escape) (.ShortRepoPath|Escape) | Str2html}}
 						{{else if eq .GetOpType 9}}
-							{{ $tagLink := .GetTag | PathEscapeSegments | Escape}}
-							{{$.i18n.Tr "action.push_tag" (.GetRepoLink|Escape) $tagLink (.ShortRepoPath|Escape) .GetTag | Str2html}}
+							{{$.i18n.Tr "action.push_tag" (.GetRepoLink|Escape) (.GetRefLink|Escape) (.ShortRepoPath|Escape) .GetTag | Str2html}}
```
Comment (jpraet) [inline templates/user/dashboard/feeds.tmpl]: The tag name and repo name are reversed:
`push_tag = pushed tag <a href="%[2]s">%[3]s</a> to <a href="%[1]s">%[4]s</a>`

Comment (zeripath) [reply templates/user/dashboard/feeds.tmpl]:
```suggestion
							{{$.i18n.Tr "action.push_tag" (.GetRepoLink|Escape) (.GetRefLink|Escape) .GetTag (.ShortRepoPath|Escape) | Str2html}}
```

File: templates/user/dashboard/feeds.tmpl [hunk 2]
```diff
@@ -17,71 +17,67 @@
...
 						{{else if eq .GetOpType 7}}
 							{{ $index := index .GetIssueInfos 0}}
-							{{$.i18n.Tr "action.create_pull_request" (.GetRepoLink|Escape) $index (.ShortRepoPath|Escape) | Str2html}}
+							{{$.i18n.Tr "action.create_pull_request" ((printf "%s/pulls/%s" .GetRepoLink $index) |Escape) $index (.ShortRepoPath|Escape) | Str2html}}
 						{{else if eq .GetOpType 8}}
 							{{$.i18n.Tr "action.transfer_repo" .GetContent (.GetRepoLink|Escape) (.ShortRepoPath|Escape) | Str2html}}
 						{{else if eq .GetOpType 9}}
-							{{ $tagLink := .GetTag | PathEscapeSegments | Escape}}
-							{{$.i18n.Tr "action.push_tag" (.GetRepoLink|Escape) $tagLink (.ShortRepoPath|Escape) .GetTag | Str2html}}
+							{{$.i18n.Tr "action.push_tag" (.GetRepoLink|Escape) (.GetRefLink|Escape) .GetTag (.ShortRepoPath|Escape) | Str2html}}
```
Comment (zeripath) [inline templates/user/dashboard/feeds.tmpl]:
```suggestion
							{{$.i18n.Tr "action.push_tag" (.GetRepoLink|Escape) (.GetRefLink|Escape) (.GetTag|Escape) (.ShortRepoPath|Escape) | Str2html}}
```

File: templates/user/dashboard/feeds.tmpl [hunk 3]
```diff
@@ -17,71 +17,67 @@
...
+							{{$.i18n.Tr "action.reject_pull_request" ((printf "%s/pulls/%s" .GetRepoLink $index) |Escape) $index (.ShortRepoPath|Escape) | Str2html}}
 						{{else if eq .GetOpType 23}}
 							{{ $index := index .GetIssueInfos 0}}
-							{{$.i18n.Tr "action.comment_pull" (.GetRepoLink|Escape) $index (.ShortRepoPath|Escape) | Str2html}}
+							{{$.i18n.Tr "action.comment_pull" ((printf "%s/pulls/%s" .GetRepoLink $index) |Escape) $index (.ShortRepoPath|Escape) | Str2html}}
 						{{else if eq .GetOpType 24}}
-							{{ $branchLink := .GetBranch | PathEscapeSegments | Escape}}
 							{{ $linkText := .Content | RenderEmoji }}
-							{{$.i18n.Tr "action.publish_release" (.GetRepoLink|Escape) $branchLink (.ShortRepoPath|Escape) $linkText | Str2html}}
+							{{$.i18n.Tr "action.publish_release" (.GetRepoLink|Escape) (.GetRefLink|Escape) (.ShortRepoPath|Escape) $linkText | Str2html}}
```
Comment (KN4CK3R) [inline templates/user/dashboard/feeds.tmpl]: This one is still wrong. Old translation was `%s/releases/tag/%s`.

Comment (zeripath) [reply templates/user/dashboard/feeds.tmpl]: You know you can make suggestions...

**Review by Gusted — APPROVED** [review]
Weird behavior for `cat-file` to not fail (almost sounds like a bug). Let's hope that other commands won't show this behavior as well.

**Comment by ryanfitzsimon** [comment]
Hi @zeripath, could you please explain how this resolves #14734?

I can't see how these changes will help in cases of slow performance, apart from in specific cases where hangs are occurring (which isn't what was reported).
The original report described PR opening being slow, but it eventually succeeded after nearly 2 minutes.

I have a repository that suffers from the performance problems, taking minutes to open PRs with large change counts.
I've just tried running `git cat-file --batch` works as expected in the bare repo on the Gitea host. I also ran a `git fsck`. Both ran without issue.

**Comment by ryanfitzsimon** [comment]
Also, what is meant by "not a valid repository"? Directories that aren't part of a repository (no `.git` directory)? Or repositories that have been somehow corrupted?

The former doesn't seem to cause a hang for me:
```console
gitserver:~$ git version
git version 2.31.1
gitserver:~$ mkdir not-a-repo
gitserver:~$ cd not-a-repo/
gitserver::~/not-a-repo$ git cat-file --batch
fatal: not a git repository (or any of the parent directories): .git
```

If it's specific to a certain type of corrupt repository, I'm not sure how to test it.

**Comment by lunny** [comment]
@ryanfitzsimon This PR should resolve part of the hang, so I have reopened #14734

**Comment by zeripath** [comment]
Ok. I had thought that we had finally found the reason for this problem but clearly you disagree. Could you update the issue and report back if things have changed on main at all?

**Pull request merged**